File: packages/wire/presence_test.go
package wire

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

type PresenceTestVector struct {
	Name             string   `json:"name"`
	KPairHex         string   `json:"k_pair_hex"`
	Timestamp        string   `json:"timestamp"`
	EpochIndex       int64    `json:"epoch_index"`
	RendezvousHandle string   `json:"rendezvous_handle"`
	SkewHandles      []string `json:"skew_handles"`
	PresenceNonceHex string   `json:"presence_nonce_hex"`
	PresenceProofHex string   `json:"presence_proof_hex"`
	BeaconNonceHex   string   `json:"beacon_nonce_hex"`
	BeaconTagHex     string   `json:"beacon_tag_hex"`
}

func TestPresenceAndLanBeaconVectors(t *testing.T) {
	kPair := sha256.Sum256([]byte("shared-k-pair-presence-vectors-v15"))
	fixedTime, _ := time.Parse(time.RFC3339, "2026-08-21T12:00:00Z")
	window := 15 * time.Minute
	epochIndex := fixedTime.Unix() / int64(window.Seconds())

	handle := DeriveRendezvousHandle(kPair[:], epochIndex)
	skewHandles := DeriveRendezvousHandlesWithSkew(kPair[:], fixedTime, window)

	presenceNonce := sha256.Sum256([]byte("fixed-presence-nonce-1"))
	proof := DerivePresenceProof(kPair[:], handle, presenceNonce[:])

	if !VerifyPresenceProof(kPair[:], handle, presenceNonce[:], proof) {
		t.Fatalf("VerifyPresenceProof failed")
	}

	beaconNonce := sha256.Sum256([]byte("fixed-beacon-nonce-1"))
	beaconTag := DeriveLanBeaconTag(kPair[:], beaconNonce[:16], epochIndex)

	vector := PresenceTestVector{
		Name:             "alice_bob_presence_and_beacon",
		KPairHex:         hex.EncodeToString(kPair[:]),
		Timestamp:        fixedTime.Format(time.RFC3339),
		EpochIndex:       epochIndex,
		RendezvousHandle: handle,
		SkewHandles:      skewHandles,
		PresenceNonceHex: hex.EncodeToString(presenceNonce[:]),
		PresenceProofHex: proof,
		BeaconNonceHex:   hex.EncodeToString(beaconNonce[:16]),
		BeaconTagHex:     hex.EncodeToString(beaconTag),
	}

	vecData, err := json.MarshalIndent([]PresenceTestVector{vector}, "", "  ")
	if err != nil {
		t.Fatalf("marshal test vector: %v", err)
	}

	targetPath := filepath.Join("testdata", "presence-vectors.json")
	if err := os.WriteFile(targetPath, vecData, 0644); err != nil {
		t.Fatalf("write presence-vectors.json: %v", err)
	}
}

func TestPresenceHandleSkewMatching(t *testing.T) {
	kPair := sha256.Sum256([]byte("shared-k-pair-presence-skew"))
	now := time.Now().UTC()
	window := 15 * time.Minute

	handleNow := DeriveRendezvousHandleForTime(kPair[:], now, window)
	if !MatchRendezvousHandle(kPair[:], handleNow, now, window) {
		t.Errorf("handleNow did not match current time")
	}

	// Boundary test: time 10 minutes earlier (within same or adjacent epoch)
	timePast := now.Add(-10 * time.Minute)
	if !MatchRendezvousHandle(kPair[:], handleNow, timePast, window) {
		t.Errorf("handleNow should match within 10 min window")
	}

	// Wrong handle should not match
	wrongKPair := sha256.Sum256([]byte("wrong-k-pair"))
	wrongHandle := DeriveRendezvousHandleForTime(wrongKPair[:], now, window)
	if MatchRendezvousHandle(kPair[:], wrongHandle, now, window) {
		t.Errorf("wrongHandle should not match")
	}
}

File: packages/protocol/src/presence.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it } from 'vitest';
import { bytesToHex, hexToBytes } from './bytes.js';
import {
  DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
  deriveLanBeaconTag,
  derivePresenceProof,
  deriveRendezvousHandle,
  deriveRendezvousHandlesWithSkew,
  matchLanBeaconTag,
  matchRendezvousHandle,
  validateRendezvousHandle,
  verifyPresenceProof,
  PresenceCoordinator,
} from './presence.js';
import { MemoryTrustStore, defaultTrustPolicy } from './trust-store.js';
import { MemorySecretResolver } from './indexeddb-secret-store.js';
import { createDeviceIdentityFromSeed } from './identity.js';

interface PresenceVector {
  name: string;
  k_pair_hex: string;
  timestamp: string;
  epoch_index: number;
  rendezvous_handle: string;
  skew_handles: string[];
  presence_nonce_hex: string;
  presence_proof_hex: string;
  beacon_nonce_hex: string;
  beacon_tag_hex: string;
}

describe('Presence and LAN discovery cross-language vector validation (sendbeam/2)', () => {
  const vectorPath = resolve(__dirname, '../../wire/testdata/presence-vectors.json');
  const vectorContent = readFileSync(vectorPath, 'utf-8');
  const vectors: PresenceVector[] = JSON.parse(vectorContent);

  it('matches Go generated presence vector byte-for-byte', async () => {
    for (const vec of vectors) {
      const kPair = hexToBytes(vec.k_pair_hex);
      const presenceNonce = hexToBytes(vec.presence_nonce_hex);
      const beaconNonce = hexToBytes(vec.beacon_nonce_hex);

      // Derive rendezvous handle
      const handle = await deriveRendezvousHandle(kPair, vec.epoch_index);
      expect(handle).toBe(vec.rendezvous_handle);

      // Derive skew handles
      const date = new Date(vec.timestamp);
      const skewHandles = await deriveRendezvousHandlesWithSkew(
        kPair,
        date,
        DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
      );
      expect(skewHandles).toEqual(vec.skew_handles);

      // Presence proof
      const proof = await derivePresenceProof(kPair, handle, presenceNonce);
      expect(proof).toBe(vec.presence_proof_hex);
      expect(await verifyPresenceProof(kPair, handle, presenceNonce, proof)).toBe(true);

      // Blinded LAN beacon tag
      const tag = await deriveLanBeaconTag(kPair, beaconNonce, vec.epoch_index);
      expect(bytesToHex(tag)).toBe(vec.beacon_tag_hex);

      // Match beacon tag
      const matched = await matchLanBeaconTag(
        kPair,
        beaconNonce,
        tag,
        date.getTime(),
        date.getTime(),
        DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS,
      );
      expect(matched).toBe(true);
    }
  });

  it('rejects stale beacon tags and invalid proofs', async () => {
    const kPair = hexToBytes('8b4c642283597de370a8313836bcc86ca6718f0d71fa4f301134d3f049da2848');
    const nonce = hexToBytes('c83888c50dc690cbf88691633df49122');
    const now = Date.now();
    const tag = await deriveLanBeaconTag(
      kPair,
      nonce,
      Math.floor(now / DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS),
    );

    // Stale timestamp (> 2 windows)
    const staleTime = now + 4 * DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS;
    const matchedStale = await matchLanBeaconTag(kPair, nonce, tag, now, staleTime);
    expect(matchedStale).toBe(false);

    // Wrong proof
    const proofValid = await verifyPresenceProof(
      kPair,
      'handle-1',
      nonce,
      bytesToHex(new Uint8Array(32)),
    );
    expect(proofValid).toBe(false);
  });

  describe('validateRendezvousHandle', () => {
    it('accepts valid 64-char lowercase hex handles', () => {
      expect(
        validateRendezvousHandle(
          '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
        ),
      ).toBe(true);
    });

    it('rejects invalid handles', () => {
      expect(validateRendezvousHandle('')).toBe(false);
      expect(validateRendezvousHandle('short')).toBe(false);
      expect(
        validateRendezvousHandle(
          '0123456789ABCDEF0123456789abcdef0123456789abcdef0123456789abcdef',
        ),
      ).toBe(false); // uppercase
      expect(
        validateRendezvousHandle(
          '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdeg',
        ),
      ).toBe(false); // non-hex
      expect(
        validateRendezvousHandle(
          '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef00',
        ),
      ).toBe(false); // too long
    });
  });

  describe('matchRendezvousHandle', () => {
    it('matches handles within skew window and rejects outside window', async () => {
      const kPair = hexToBytes('8b4c642283597de370a8313836bcc86ca6718f0d71fa4f301134d3f049da2848');
      const now = 1700000000000;
      const epochIndex = Math.floor(now / DEFAULT_RENDEZVOUS_EPOCH_WINDOW_MS);

      const currentHandle = await deriveRendezvousHandle(kPair, epochIndex);
      const prevHandle = await deriveRendezvousHandle(kPair, epochIndex - 1);
      const nextHandle = await deriveRendezvousHandle(kPair, epochIndex + 1);
      const oldHandle = await deriveRendezvousHandle(kPair, epochIndex - 5);

      expect(await matchRendezvousHandle(kPair, currentHandle, now)).toBe(true);
      expect(await matchRendezvousHandle(kPair, prevHandle, now)).toBe(true);
      expect(await matchRendezvousHandle(kPair, nextHandle, now)).toBe(true);
      expect(await matchRendezvousHandle(kPair, oldHandle, now)).toBe(false);
    });
  });

  describe('PresenceCoordinator and truthful presence', () => {
    it('manages truthful presence lifecycle and enforces epoch expiry', async () => {
      const trustStore = new MemoryTrustStore();
      const secretResolver = new MemorySecretResolver();
      const coordinator = new PresenceCoordinator(trustStore, secretResolver, 30_000);

      const seed = hexToBytes('4444444444444444444444444444444444444444444444444444444444444444');
      const id = await createDeviceIdentityFromSeed(seed);
      const deviceId = id.deviceId;
      const kPair = hexToBytes('8b4c642283597de370a8313836bcc86ca6718f0d71fa4f301134d3f049da2848');
      const credRef = 'cred_1';

      await trustStore.addOrUpdateDevice({
        deviceId,
        localLabel: 'Peer 1',
        publicKey: bytesToHex(id.publicKey),
        pairCredentialRef: credRef,
        pairingRole: 'offerer',
        capabilities: ['transfer.v1'],
        trustedAt: new Date().toISOString(),
        firstSeenAt: new Date().toISOString(),
        policy: defaultTrustPolicy(),
      });
      secretResolver.setSecret(deviceId, kPair);

      const t0 = 1700000000000;
      const activeHandlesMap = await coordinator.getActiveHandles(t0);
      const activeHandles = activeHandlesMap.get(deviceId)!;
      expect(activeHandles).toBeDefined();
      expect(activeHandles.length).toBe(3);

      // Initially peer is offline
      expect(coordinator.getPeerPresence(deviceId, t0)).toBe('offline');

      // Match incoming handle
      const match = await coordinator.matchInboundPresence(
        activeHandles[0]!,
        undefined,
        undefined,
        t0,
      );
      expect(match).not.toBeNull();
      expect(match!.deviceId).toBe(deviceId);

      // Status should now be 'discovered'
      expect(coordinator.getPeerPresence(deviceId, t0)).toBe('discovered');

      // Check within epoch window (t0 + 15s)
      expect(coordinator.getPeerPresence(deviceId, t0 + 15_000)).toBe('discovered');

      // Truthful presence: once epoch window expires (t0 + 31s), status must be 'offline'
      expect(coordinator.getPeerPresence(deviceId, t0 + 31_000)).toBe('offline');

      // Transition to connecting
      coordinator.setPeerConnecting(deviceId);
      expect(coordinator.getPeerPresence(deviceId, t0 + 31_000)).toBe('connecting');

      // Transition to connected
      coordinator.setPeerConnected(deviceId);
      expect(coordinator.getPeerPresence(deviceId, t0 + 31_000)).toBe('connected');

      // Transition to disconnected
      coordinator.setPeerDisconnected(deviceId);
      expect(coordinator.getPeerPresence(deviceId, t0 + 31_000)).toBe('offline');

      // Prune expired
      coordinator.recordPeerDiscovered(deviceId, activeHandles[0]!, t0);
      coordinator.pruneExpired(t0 + 31_000);
      expect(coordinator.getPeerPresence(deviceId, t0 + 31_000)).toBe('offline');
    });
  });
});
